building-and-deploying-a-fastapi-app.md

```diff
@@ -0,0 +1,541 @@
+
+In `app/model.py`, write the following:
+
+```
+from pydantic import BaseModel, Field
+from typing import Optional, List
+
+
+class RecipeSchema(BaseModel):
+    id: Optional[int]
+    name: str = Field(...)
+    ingredients: List[str] = Field(...)
+
+    class Config:
+        schema_extra = {
+            "example": {
+                "name": "Donuts",
+                "ingredients": ["Flour", "Milk", "Sugar", "Vegetable Oil"]
+            }
+        }
+
+
+class UpdateRecipeSchema(BaseModel):
+    name: Optional[str]
+    ingredients: Optional[List[str]]
+
+    class Config:
+        schema_extra = {
+            "example": {
+                "name": "Buns",
+                "ingredients": ["Flour", "Milk", "Sugar", "Vegetable Oil"]
+            }
+        }
+
+
+```
+
+In the block of code above, you defined how the recipe would be represented, sent, and retrieved in your in-app database. Each recipe will have an ID automatically generated once a POST request is sent. Each recipe will also comprise of a name and an array of ingredients. If any other type of data not listed in the schema is sent in the request body, an error will be returned.
+
+The `RecipeSchema` has a subclass `Config`. The subclass contains an object variable, `schema_extra`, which includes a key `example` used as the mock data in the interactive documentation.
+
+The `UpdateRecipeSchema`, on the other hand, is the model for guiding request body sent through the UPDATE route. The difference between this model and the `RecipeSchema` is that the schema values can be passed optionally.
+
+With the model in place, let's write the code for the application's routes. In `app/api.py`, add the recipes database before the base route __"/"__:
+
+```py
+recipes = [
+    {
+        "id": 1,
+        "name": "Donuts",
+        "ingredients": ["Flour", "Milk", "Sugar", "Vegetable Oil"]
+    }
+]
+```
+
+Below the base route, add the GET routes:
+
+```
+@app.get("/recipe", tags=["Recipe"])
+def get_recipes() -> dict:
+    return {
+        "data": recipes
+    }
+
+@app.get("/recipe/{id}", tags=["Recipe"])
+def get_recipe(id: int) -> dict:
+    if id > len(recipes) or id < 1:
+        return {
+            "error": "Invalid ID passed."
+        }
+
+    for recipe in recipes:
+        if recipe['id'] == id:
+            return {
+                "data": [
+                    recipe
+                ]
+            }
+
+    return {
+        "error": "No such recipe with ID {} exist".format(id)
+    }
+```
+
+You defined GET routes for retrieving all the recipes and a single recipe using its ID in the code block above. An error message is returned if an invalid ID is passed as a parameter in the __/recipe/{id}__ route. Test the routes by visiting http://localhost:8080/recipe and http://localhost:8080/recipe/1:
+
+![Recipes](https://res.cloudinary.com/adeshina/image/upload/v1608635876/zree0awqm8co1g2kf339.png)
```

Comment:
could you include the images here, so we can put them in the CMS instead of hot-linking them?

Reply:
Oh, okay. 


Is this better?
1. Recipes

![](https://res.cloudinary.com/adeshina/image/upload/v1608635876/zree0awqm8co1g2kf339.png)

2. First Recipe
 ![](https://res.cloudinary.com/adeshina/image/upload/v1608635909/gdted1ywrnnkjupy9x2o.png)

3. Okteto Dashboard
![Okteto Dashboard](https://res.cloudinary.com/adeshina/image/upload/v1608635743/x9tmv4styid2hffxilxx.png)

4. Deployed app
![Deployed Application](https://res.cloudinary.com/adeshina/image/upload/v1608636581/jq7fjtldhz1tqqzuh29h.png)

5. Last dashboard image:
![Dashboard](https://res.cloudinary.com/adeshina/image/upload/v1608638743/q8xcjhl2gdnh0tdzr573.png)